### PR TITLE
Skip no closure check for non-Crystal procs

### LINKED ISSUE #7279: Module validation failed: Call parameter type does not match function signature!
Crystal 0.27.0 (2018-11-04)
LLVM: 6.0.1
Default target: x86_64-apple-macosx

Attempting to compile this code produces the following error:

```cr
lib LibFoo
  struct S
    callback : ->
  end
end

s = LibFoo::S.new
s.callback = nil
```

```
Module validation failed: Call parameter type does not match function signature!
%Nil ()* null
 %"->" = type { i8*, i8* }  %71 = call i8* @"~check_proc_is_not_closure"(%Nil ()* null), !dbg !74
 (Exception)
  from Crystal::CodeGenVisitor#finish:Nil
  from Crystal::Compiler#codegen<Crystal::Program, Crystal::ASTNode+, Array(Crystal::Compiler::Source), String>:(Tuple(Array(Crystal::Compiler::CompilationUnit), Array(String)) | Nil)
  from Crystal::Compiler#compile<Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result
  from Crystal::Command#run_command<Bool>:Nil
  from Crystal::Command#run:(Bool | Crystal::Compiler::Result | Nil)
  from __crystal_main
  from main
```

### PR DESCRIPTION
**Pull request by jhass**
They can't have a closure anyways and more importantly are represented as a
single pointer. So the check could not even deconstruct them into two
pointers. This previously generated invalid code trying to pass the
function pointer type to the check function expecting the Crystal proc type

fixes #7279

### CODE REVIEW

File: src/compiler/crystal/codegen/codegen.cr [hunk 1]
```diff
@@ -1530,10 +1531,13 @@ module Crystal
...
     def check_proc_is_not_closure(value, type)
-      check_fun_name = "~check_proc_is_not_closure"
-      func = @main_mod.functions[check_fun_name]? || create_check_proc_is_not_closure_fun(check_fun_name)
-      func = check_main_fun check_fun_name, func
-      value = call func, [value] of LLVM::Value
+      if value.type == llvm_typer.proc_type
+        check_fun_name = "~check_proc_is_not_closure"
+        func = @main_mod.functions[check_fun_name]? || create_check_proc_is_not_closure_fun(check_fun_name)
+        func = check_main_fun check_fun_name, func
+        proc_value = bit_cast value, llvm_typer.proc_type
```
Comment (RX14) [inline src/compiler/crystal/codegen/codegen.cr:1538]: Is this extra bit case needed if we already checked `value.type`?

File: spec/compiler/codegen/closure_spec.cr [hunk 2]
```diff
@@ -681,4 +681,51 @@ describe "Code gen: closure" do
       f1.call &+ f2.call
     ))
   end
+
+  it "ensures it can raise from the closure check" do
+    expect_raises(Exception, /NoReturn/) do
```
Comment (straight-shoota) [inline spec/compiler/codegen/closure_spec.cr:686]: Can you compare with the entire expected error message here? It helps so much understanding the purpose of the spec.
Also Regex is unnecessary, String is fine.

File: src/compiler/crystal/codegen/codegen.cr (hunk 1 above)
Comment (jhass) [reply src/compiler/crystal/codegen/codegen.cr]: Yup, sorry. Totally missed that out from an earlier experiment, good catch.

**Comment by jhass** [comment]
Thanks! Would love to see (solutions for) #9253 and #9255 in 0.35 too <3

**Pull request merged**